### PR TITLE
Atomize topk reduction

### PR DESCRIPTION
**Pull request by jackgerrits**
Also fixes example copying bug.

#2003

### CODE REVIEW

File: vowpalwabbit/topk.h [hunk 1]
```diff
@@ -3,4 +3,36 @@ Copyright (c) by respective owners including Yahoo!, Microsoft, and
...
+    struct compare_scored_examples
+    {
+      constexpr bool operator()(scored_example const& a, scored_example const& b) const { return a.first > b.first; }
+    };
+
+    void update_priority_queue(float pred, v_array<char>& tag);
+
+    // rec number
+    const uint32_t m_K;
+    std::priority_queue<scored_example, std::vector<scored_example>, compare_scored_examples> m_pr_queue;
```
[the remark this answers is not in the document]
Comment (rajan-chari) [reply vowpalwabbit/topk.h]: Seems like we are only interested in the top element.  A priority queue likely more efficient.  Also it's backed std::vector by default and would require less allocations.  

But as in all things performance... measure?

File: vowpalwabbit/topk.h [hunk 2]
```diff
@@ -3,4 +3,36 @@ Copyright (c) by respective owners including Yahoo!, Microsoft, and
...
+  private:
+    struct compare_scored_examples
+    {
+      constexpr bool operator()(scored_example const& a, scored_example const& b) const { return a.first > b.first; }
+    };
+
+    void update_priority_queue(float pred, v_array<char>& tag);
+
+    // rec number
+    const uint32_t m_K;
```
Comment (rajan-chari) [inline vowpalwabbit/topk.h]: _ instead of m_?

File: vowpalwabbit/topk.h [hunk 3]
```diff
@@ -3,4 +3,36 @@ Copyright (c) by respective owners including Yahoo!, Microsoft, and
...
+{
+  struct topk
+  {
+    using scored_example = std::pair<float, v_array<char>>;
+
+    topk(uint32_t K);
+
+    void predict(LEARNER::single_learner& base, multi_ex& ec_seq);
+    void learn(LEARNER::single_learner& base, multi_ex& ec_seq);
+    std::vector<scored_example> drain_queue();
```
Comment (rajan-chari) [inline vowpalwabbit/topk.h]: Should this be private?

File: vowpalwabbit/topk.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jackgerrits) [reply vowpalwabbit/topk.h]: This reduction is not used very much, there is already a priority queue here. I think we should leave as is

File: vowpalwabbit/topk.h (hunk 3 above)
Comment (jackgerrits) [reply vowpalwabbit/topk.h]: It doesn't seem right to link driver related output functions to the class itself. For the output function to work it needs access to the result of the reduction, hence why this function exists.

File: vowpalwabbit/topk.cc [hunk 4]
```diff
@@ -5,51 +5,77 @@ license as described in the file LICENSE.
...
+  else if (m_pr_queue.top().first < pred)
+  {
+    m_pr_queue.pop();
+    m_pr_queue.push(std::make_pair(pred, tag));
+  }
+}
+
+std::vector<VW::topk::scored_example> VW::topk::drain_queue()
+{
+  std::vector<scored_example> result;
```
Comment (rajan-chari) [inline vowpalwabbit/topk.cc]: Won't this cause allocations for every finish_example?

Comment (peterychang) [reply vowpalwabbit/topk.cc]: This was the original reason I suggested the multimap. priority queue is fine for the top element, like you mentioned above, but this was bad

Comment (jackgerrits) [reply vowpalwabbit/topk.cc]: Yep when using a pr it has to copy out into a vector

**Comment by jackgerrits** [comment]
Changed to multimap, now there is no copy to output.

**Pull request merged**